File: scripts/validate.py
#!/usr/bin/env python3
"""Validate synthpunk palette files."""

import json
import os
import re


def validate_hex(hex_str):
    return re.match(r"^[0-9A-Fa-f]{6}$", hex_str) is not None


def validate_palette(path, name):
    print(f"\nValidating {name}...")
    with open(path) as f:
        data = json.load(f)
    
    colors = data.get("colors", {})
    errors = []
    
    for color_name, color_data in colors.items():
        hex_val = color_data.get("hex", "")
        rgb = color_data.get("rgb", [])
        
        if not validate_hex(hex_val):
            errors.append(f"  Invalid hex for {color_name}: {hex_val}")
        
        if len(rgb) != 3 or not all(isinstance(v, int) and 0 <= v <= 255 for v in rgb):
            errors.append(f"  Invalid RGB for {color_name}: {rgb}")
        
        # Verify hex matches rgb
        expected_hex = f"{rgb[0]:02x}{rgb[1]:02x}{rgb[2]:02x}".upper()
        if hex_val.upper() != expected_hex:
            errors.append(f"  Hex/RGB mismatch for {color_name}: hex={hex_val}, rgb={rgb}")
    
    if errors:
        print(f"  FAILED with {len(errors)} errors:")
        for e in errors:
            print(e)
        return False
    else:
        print(f"  OK — {len(colors)} colors validated")
        return True


def validate_zed_theme(path, name):
    print(f"\nValidating {name}...")
    with open(path) as f:
        data = json.load(f)
    
    themes = data.get("themes", [])
    errors = []
    
    for theme in themes:
        style = theme.get("style", {})
        for key, value in style.items():
            if key == "syntax":
                for syntax_key, syntax_value in value.items():
                    color = syntax_value.get("color", "")
                    if color and not re.match(r"^#[0-9A-Fa-f]{8}$", color):
                        errors.append(f"  Invalid color in syntax.{syntax_key}: {color}")
            elif key == "players":
                continue
            elif value and not re.match(r"^#[0-9A-Fa-f]{8}$", str(value)):
                errors.append(f"  Invalid color in {key}: {value}")
    
    if errors:
        print(f"  FAILED with {len(errors)} errors:")
        for e in errors:
            print(e)
        return False
    else:
        print(f"  OK")
        return True


def main():
    project_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

    results = []
    # Pastel palettes
    results.append(validate_palette(os.path.join(project_dir, "palette", "base.json"), "base.json (pastel light)"))
    results.append(validate_palette(os.path.join(project_dir, "palette", "dark.json"), "dark.json (pastel dark)"))
    # Neon palettes
    results.append(validate_palette(os.path.join(project_dir, "palette", "neon-light.json"), "neon-light.json"))
    results.append(validate_palette(os.path.join(project_dir, "palette", "neon-dark.json"), "neon-dark.json"))
    # Zed themes
    results.append(validate_zed_theme(os.path.join(project_dir, "themes", "zed", "themes", "synthpunk-pastel.json"), "Zed synthpunk-pastel theme"))
    results.append(validate_zed_theme(os.path.join(project_dir, "themes", "zed", "themes", "synthpunk-neon.json"), "Zed synthpunk-neon theme"))

    print("\n" + "="*50)
    if all(results):
        print("All validations passed!")
    else:
        print("Some validations failed. Please fix the errors above.")


if __name__ == "__main__":
    main()

File: generator/validate.ts
import path from "node:path";
import { loadPalette } from "./palette";
import { loadScopes, loadSyntaxMapping } from "./syntaxMapping";
import type { VariantName } from "./types";
import { loadUIMapping } from "./uiMapping";

const PROJECT_DIR = path.resolve(import.meta.dir, "../");
const PALETTE_DIR = path.join(PROJECT_DIR, "palette");

const VARIANTS: VariantName[] = [
	"pastel-dark",
	"pastel-light",
	"neon-dark",
	"neon-light",
];

function validatePalettes(): boolean {
	let ok = true;
	for (const variant of VARIANTS) {
		try {
			const palette = loadPalette(PALETTE_DIR, variant);
			const colorNames = Object.keys(palette.colors);
			let errors = 0;
			for (const name of colorNames) {
				const color = palette.colors[name];
				const expectedHex = `${color.rgb[0].toString(16).padStart(2, "0").toUpperCase()}${color.rgb[1].toString(16).padStart(2, "0").toUpperCase()}${color.rgb[2].toString(16).padStart(2, "0").toUpperCase()}`;
				if (color.hex.toUpperCase() !== expectedHex) {
					console.error(
						`  Hex/RGB mismatch for ${name}: hex=${color.hex}, expected=${expectedHex}`,
					);
					errors++;
				}
			}
			if (errors > 0) {
				ok = false;
			} else {
				console.log(
					`  OK — ${palette.info.name}: ${colorNames.length} colors validated`,
				);
			}
		} catch (e) {
			console.error(`  FAILED: ${variant}: ${e}`);
			ok = false;
		}
	}
	return ok;
}

function validateMappings(): boolean {
	let ok = true;
	const uiMapping = loadUIMapping(PALETTE_DIR);
	const syntaxMapping = loadSyntaxMapping(PALETTE_DIR);

	for (const variant of VARIANTS) {
		const palette = loadPalette(PALETTE_DIR, variant);

		const groups = [
			"background",
			"surface",
			"text",
			"border",
			"editor",
			"terminal",
			"status",
			"vcs",
		] as const;
		for (const group of groups) {
			const groupMap = (
				uiMapping as unknown as Record<string, Record<string, string>>
			)[group];
			for (const [key, colorName] of Object.entries(groupMap)) {
				if (!(colorName in palette.colors)) {
					console.error(
						`  Unknown color "${colorName}" in ui-mapping.${group}.${key} for ${variant}`,
					);
					ok = false;
				}
			}
		}

		const syntaxGroups = [
			"keywords",
			"storage",
			"literals",
			"comments",
			"functions",
			"variables",
			"entities",
			"support",
			"invalid",
			"emphasis",
		] as const;
		for (const group of syntaxGroups) {
			const groupMap = (
				syntaxMapping as unknown as Record<string, Record<string, string>>
			)[group];
			if (!groupMap) continue;
			for (const [key, colorName] of Object.entries(groupMap)) {
				if (!(colorName in palette.colors)) {
					console.error(
						`  Unknown color "${colorName}" in syntax.${group}.${key} for ${variant}`,
					);
					ok = false;
				}
			}
		}

		console.log(`  OK — ${variant} mapping validations passed`);
	}
	return ok;
}

function validateScopes(): boolean {
	const scopes = loadScopes(PALETTE_DIR);
	const syntaxMapping = loadSyntaxMapping(PALETTE_DIR);
	let ok = true;
	for (const entry of scopes) {
		const parts = entry.syntaxRole.split(".");
		const group = parts[0];
		const role = parts.slice(1).join(".");
		const groupMap = (
			syntaxMapping as unknown as Record<string, Record<string, string>>
		)[group];
		if (!groupMap || !(role in groupMap)) {
			console.error(
				`  Unknown syntaxRole "${entry.syntaxRole}" in scope "${entry.name}"`,
			);
			ok = false;
		}
	}
	if (ok) console.log(`  OK — ${scopes.length} scopes validated`);
	return ok;
}

export function validate(): boolean {
	console.log("Validating palettes...");
	const p = validatePalettes();
	console.log("\nValidating mappings...");
	const m = validateMappings();
	console.log("\nValidating scopes...");
	const s = validateScopes();
	console.log(`\n${"=".repeat(50)}`);
	if (p && m && s) {
		console.log("All validations passed!");
		return true;
	}
	console.log("Some validations failed.");
	return false;
}

if (import.meta.main) {
	const success = validate();
	process.exit(success ? 0 : 1);
}
